File: arrays/quick-sort/t.js
//divide and conquer

function quickSort(arr){
    if(arr.length<2) return arr
    let pivot=arr[arr.length-1]
    let left=[]
    let right=[]
    for(let i=0;i<arr.length-1;i++){
        if(arr[i]<pivot) left.push(arr[i])
        else right.push(arr[i])
    }
    return [...quickSort(left),pivot,...quickSort(right)];
}
/**
 * average time complexity O(nlogn) // array roughly divides into two halves at each step
 * worst time complexity O(n^2), pivot is always smallest ot largest
 * space complexity : creates 2 new arrays O(n^2) recursion stack takes O(log n) space in average case and O(n) in the worst case
 */
// console.log(quickSort([4,56,21,46,78,11,25]));

// in place quick sort without extra space
  function quickSortInPlace(arr,low=0,high=arr.length-1){
    if(low>=high){
      return arr
    }
    const partitionIndex = partitionIndexing(arr,low,high)
    quickSortInPlace(arr,low,partitionIndex-1)
    quickSortInPlace(arr,partitionIndex+1,high)
    return arr
  }
  function partitionIndexing(arr,low,high){
    let i =low-1
    let pivot = arr[high]
    for(let j=low;j<high;j++){
      if(arr[j]<pivot){
        i++;
        [arr[i],arr[j]]=[arr[j],arr[i]]
      }
    }
    [arr[i+1],arr[high]]=[arr[high],arr[i+1]]
    return i+1
  }
  const myArray = [10, 7, 8, 9, 1, 5];
  console.log(quickSortInPlace(myArray))
/**
 * avg tc O(nlogn) - same reasoning as above
 * worst tc O(n^2) - same ...
 * sc - auxiliary space O(logn) coz only space needed for recursion stack
 */

File: arrays/quick-sort/t.py
def quick_sort(arr):
    if len(arr)<2: return arr
    pivot = arr[-1]
    left=[]
    right=[]
    for i in range(len(arr)-1):
        if(arr[i]<pivot):
            left.append(arr[i])
        else:
            right.append(arr[i])
    return quick_sort(left)+[pivot]+quick_sort(right)
# print(quick_sort([4,56,21,46,78,11,25]))

def quick_sort_in_place(arr,low=0,high=None):
    if high is None: high = len(arr) -1
    if(low>=high): return arr
    partition_index = partitioning(arr,low,high)
    quick_sort_in_place(arr,low,partition_index-1)
    quick_sort_in_place(arr,partition_index+1,high)
    return arr
def partitioning(arr,low,high):
    i = low-1
    pivot = arr[high]
    for j in range(low,high):
        if(arr[j]<pivot):
            i+=1
            arr[i],arr[j]=arr[j],arr[i]
    arr[i+1],arr[high]=arr[high],arr[i+1]
    return i+1
print(quick_sort_in_place([4,56,21,46,78,11,25]))